File: library/foundation/libfoundation/cpp_foundation/cqlog.cpp
#include "cqlog.hh"
#include "cqlog_p.h"
#include <cstdarg>
#include <cstdio>

void cqLogInfo(const char *file, int line, const char *format, ...) {
    char message[256]; {
        va_list args;
        va_start(args, format);
        vsnprintf(message, sizeof(message), format, args);
        va_end(args);
    }
    cq_log_info(file, line, message);
}

void cqLogError(const char *file, int line, const char *format, ...) {
    char message[256]; {
        va_list args;
        va_start(args, format);
        vsnprintf(message, sizeof(message), format, args);
        va_end(args);
    }
    cq_log_error(file, line, message);
}

File: library/foundation/MacFoundation/CQLog.m
#import "CQLog.h"
#import "cqfoundationarche.h"

static void logging(NSString *tag, NSString *file, int line, NSString *format, va_list args) {
    if (format.length == 0) {
        return;
    }
    
    //[NSString initWithFormat:arguments:] depends on local text encoding,
    //use vsnprintf() to splice utf-8 string.
    char buffer[1024 * 4];
    vsnprintf(buffer, sizeof(buffer), format.UTF8String, args);
    NSString *message = [NSString stringWithUTF8String:buffer];
    
    if (file.length > 0 && line > 0) {
        NSLog(@"%@|%@(%04d)|%@", tag, file.lastPathComponent, line, message);
    } else {
        NSLog(@"%@|%@", tag, message);
    }
}

void CQLogInfo(NSString *file, int line, NSString *format, ...) {
    va_list args;
    va_start(args, format);
    logging(@"I", file, line, format, args);
    va_end(args);
}

void CQLogError(NSString *file, int line, NSString *format, ...) {
    va_list args;
    va_start(args, format);
    logging(@"E", file, line, format, args);
    va_end(args);
}
